codec/address/bech32_codec_test.go

```diff
@@ -0,0 +1,156 @@
+	for i := 0; i < totalAddresses; i++ {
+		_, err := rand.Read(addr)
+		if err != nil {
+			return nil, err
+		}
+		keys[i] = addr
+	}
+
+	return keys, nil
+}
+
+func TestNewBech32Codec(t *testing.T) {
+	tests := []struct {
+		name    string
+		prefix  string
+		lru     *simplelru.LRU
+		address string
+	}{
+		{
+			name:    "create accounts cached bech32 codec",
+			prefix:  "cosmos",
+			lru:     accAddrCache,
+			address: "cosmos1p8s0p6gqc6c9gt77lgr2qqujz49huhu6a80smx",
+		},
+		{
+			name:    "create validator cached bech32 codec",
+			prefix:  "cosmosvaloper",
+			lru:     valAddrCache,
+			address: "cosmosvaloper1sjllsnramtg3ewxqwwrwjxfgc4n4ef9u2lcnj0",
+		},
+		{
+			name:    "create consensus cached bech32 codec",
+			prefix:  "cosmosvalcons",
+			lru:     consAddrCache,
+			address: "cosmosvalcons1ntk8eualewuprz0gamh8hnvcem2nrcdsgz563h",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.Equal(t, tt.lru.Len(), 0)
+			ac := NewBech32Codec(tt.prefix)
+			cached, ok := ac.(cachedBech32Codec)
+			assert.Assert(t, ok)
+			assert.Equal(t, cached.cache, tt.lru)
+
+			addr, err := ac.StringToBytes(tt.address)
+			assert.NilError(t, err)
+			assert.Equal(t, tt.lru.Len(), 1)
+
+			cachedAddr, ok := tt.lru.Get(tt.address)
+			assert.Assert(t, ok)
+			assert.DeepEqual(t, addr, cachedAddr)
+
+			accAddr, err := ac.BytesToString(addr)
+			assert.NilError(t, err)
+			assert.Equal(t, tt.lru.Len(), 2)
+
+			cachedStrAddr, ok := tt.lru.Get(cached.codec.Bech32Prefix + conv.UnsafeBytesToStr(addr))
+			assert.Assert(t, ok)
+			assert.DeepEqual(t, accAddr, cachedStrAddr)
+		})
+	}
+}
+
+func TestMultipleBech32Codec(t *testing.T) {
+	cosmosAc, ok := NewBech32Codec("cosmos").(cachedBech32Codec)
+	assert.Assert(t, ok)
+	stakeAc := NewBech32Codec("stake").(cachedBech32Codec)
+	assert.Assert(t, ok)
+	assert.Equal(t, cosmosAc.cache, stakeAc.cache)
+
+	addr := make([]byte, 32)
+	_, err := rand.Read(addr)
+	assert.NilError(t, err)
+
+	cosmosAddr, err := cosmosAc.BytesToString(addr)
+	assert.NilError(t, err)
+	stakeAddr, err := stakeAc.BytesToString(addr)
+	assert.NilError(t, err)
+	assert.Assert(t, cosmosAddr != stakeAddr)
+
+	cachedCosmosAddr, err := cosmosAc.BytesToString(addr)
+	assert.NilError(t, err)
+	assert.Equal(t, cosmosAddr, cachedCosmosAddr)
+
+	cachedStakeAddr, err := stakeAc.BytesToString(addr)
+	assert.NilError(t, err)
+	assert.Equal(t, stakeAddr, cachedStakeAddr)
+}
+
+func TestBech32CodecRace(t *testing.T) {
+	ac := NewBech32Codec("cosmos")
+
+	workers := 4
+	done := make(chan bool, workers)
+	cancel := make(chan bool)
+
+	for i := byte(1); i <= 2; i++ { // works which will loop in first 100 addresses
+		go bytesToStringCaller(t, ac, i, 100, cancel, done)
+	}
+
+	for i := byte(1); i <= 2; i++ { // works which will generate 1e6 new addresses
+		go bytesToStringCaller(t, ac, i, 1000000, cancel, done)
+	}
+
+	<-time.After(time.Millisecond * 30)
+	close(cancel)
+
+	// cleanup
+	for i := 0; i < 4; i++ {
+		<-done
```

Comment:
🤔 this condition does not ensure that a cache entry was written. It may block for the whole time  and still pass. The workers execute in parallel but they likely work on different keys. You can have stronger checks if you refactor this to sync execution. For example:
```
func TestBech32CodecRace2(t *testing.T) {
	ac := NewBech32Codec("cosmos")
	myAddrBz := []byte{0x1, 0x2, 0x3, 0x4, 0x5}

	var (
		wgStart, wgDone sync.WaitGroup
		errCount        atomic.Uint32
	)
	const n = 3
	wgStart.Add(n)
	wgDone.Add(n)
	for i := 0; i < n; i++ {
		go func() {
			wgStart.Done()
			wgStart.Wait() // wait for all routines started

			got, err := ac.BytesToString(myAddrBz)
			if err != nil || got != "cosmos1qypqxpq9dc9msf" {
				errCount.Add(1)
			}
			wgDone.Done()
		}()
	}
	wgDone.Wait() // wait for all routines completed
	assert.Equal(t, errCount.Load(), uint32(0))
}
```

Reply:
Nice catch just updated it